src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -105,32 +154,43 @@ object CuratorElectionStream extends StrictLogging {
      * initial leader node asynchronously. If we poll for leader information before this background hook completes, then
      * a KeeperException.NoNodeException is thrown (which we handle, and retry)
      */
-    def longPollLeaderChange(retries: Int = 0): Unit = singleThreadEC.execute { () =>
-      try {
-        if (latch.getState == LeaderLatch.State.STARTED)
+    def longPollLeaderChange(id: Int, retries: Int = 0): Unit = singleThreadEC.execute { () =>
+      if (currentLoopId.get() != id) {
+        logger.info(s"Leader watch loop ${id} cancelled")
+      } else {
+        try {
+          logger.debug(s"Leader watch loop ${id} registering watch")
           client.getChildren
             .usingWatcher(new CuratorWatcher {
-              override def process(event: WatchedEvent): Unit =
-                if (!_isCancelled) longPollLeaderChange()
+              override def process(event: WatchedEvent): Unit = {
+                logger.debug(s"Leader watch loop ${id} event received: ${event}")
+                if (event.getType != EventType.None) // ignore connection errors
+                  longPollLeaderChange(id)
+              }
             })
             .forPath(curatorLeaderLatchPath)
-        emitLeader()
-      } catch {
-        case ex: KeeperException.NoNodeException if retries < 100 =>
-          // Wait for node to be created
-          logger.info("retrying")
-          Thread.sleep(retries * 10L)
-          longPollLeaderChange(retries + 1)
-        case ex: Throwable =>
-          sq.fail(ex)
+          emitLeader(id)
+        } catch {
+          case ex: KeeperException.NoNodeException if retries < 10 =>
```

Comment:
Was it on purpose that `retries` went from 100 down to 10?

Reply:
- [ ] Yes, I'll add a comment